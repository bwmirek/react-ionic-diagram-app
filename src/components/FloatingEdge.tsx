import { EdgeProps, useNodes } from 'react-flow-renderer';
import calculateNodesIntersectionPosition from '../utils/calculateNodesIntersectionPosition';

function FloatingEdge(props: EdgeProps) {
  const nodes = useNodes();
  const sourceNode = nodes.find((node) => node.id === props.source);
  const targetNode = nodes.find((node) => node.id === props.target);

  if (!sourceNode || !targetNode) {
    return null;
  }

  const sourceIntersectionPoint = calculateNodesIntersectionPosition(sourceNode, targetNode);
  const targetIntersectionPoint = calculateNodesIntersectionPosition(targetNode, sourceNode);

  const d = `M ${sourceIntersectionPoint.x},${sourceIntersectionPoint.y}L ${targetIntersectionPoint.x},${targetIntersectionPoint.y}`;

  return (
    <g className="react-flow__connection">
      <path id={props.id} className="react-flow__edge-path" d={d} markerEnd={props.markerEnd} style={props.style}/>
    </g>
  );
}

export default FloatingEdge;
