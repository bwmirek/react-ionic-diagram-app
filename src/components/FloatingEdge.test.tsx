import { render } from '@testing-library/react';
import { Node, Position, } from 'react-flow-renderer';
import FloatingEdge from './FloatingEdge';

const firstNode: Node = {
  id: 'node-source',
  position: {
    x: 0,
    y: 0,
  },
  data: {},
  width: 150,
  height: 36,
};

const secondNode: Node = {
  id: 'node-target',
  position: {
    x: 100,
    y: 100,
  },
  data: {},
  width: 150,
  height: 36,
};

jest.mock('react-flow-renderer', () => ({
  useNodes: () => [firstNode, secondNode]
}));

jest.mock('../utils/calculateNodesIntersectionPosition', () => () => ({ x: 100, y: 100 }));

// Fix Warning: The tag <path> is unrecognized in this browser.
// by wrapping it with <svg>
const renderElement = (sourceNode: Node, targetNode: Node, sourceId: string, targetId: string) => {
  return render(
    <svg>
      <FloatingEdge
        id={'test'}
        sourcePosition={'bottom' as Position}
        targetPosition={'top' as Position}
        source={sourceId}
        target={targetId}
        sourceX={sourceNode.position.x}
        sourceY={sourceNode.position.y}
        targetX={targetNode.position.x}
        targetY={targetNode.position.y}
      />
    </svg>
  );
};

describe('FloatingEdge', () => {
  it('should render without crashing', () => {
    const { baseElement } = renderElement(firstNode, secondNode, firstNode.id, secondNode.id);

    expect(baseElement).toMatchSnapshot();
  });
  it('should render empty', () => {
    const { baseElement } = renderElement(firstNode, secondNode, firstNode.id, 'node-missing');

    expect(baseElement).toMatchSnapshot();
  });
});

