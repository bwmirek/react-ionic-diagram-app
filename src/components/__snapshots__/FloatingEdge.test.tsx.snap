// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatingEdge should render empty 1`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`FloatingEdge should render without crashing 1`] = `
<body>
  <div>
    <svg>
      <g
        class="react-flow__connection"
      >
        <path
          class="react-flow__edge-path"
          d="M 100,100L 100,100"
          id="test"
        />
      </g>
    </svg>
  </div>
</body>
`;
